tests/test_transform.py

```diff
@@ -0,0 +1,182 @@
+        centered_eight,
+        transform_funcs=funcs,
+        group="posterior_predictive",
+    )
+
+    # return group is only group in idata
+    assert "/unconstrained_posterior_predictive" in idata.groups
+
+    dt_out = idata["unconstrained_posterior_predictive"]
+
+    # y transformed
+    # expected values of y
+    expected_y = centered_eight["posterior_predictive"]["y"] - 3
+    xrt.assert_equal(dt_out["y"], expected_y)
+
+
+def test_with_transform_return_dataset(centered_eight):
+    # transform mu
+    funcs = {"mu": lambda arr: arr - 3}
+
+    ds_uc = get_unconstrained_samples(
+        centered_eight,
+        transform_funcs=funcs,
+        return_dataset=True,
+    )
+
+    # it returns a datatree
+    assert isinstance(ds_uc, xr.Dataset)
+
+    # y transformed
+    # expected values of mu
+    expected_mu = centered_eight["posterior"]["mu"] - 3
+    xrt.assert_equal(ds_uc["mu"], expected_mu)
+
+
+def test_with_transform_filter_vars(centered_eight):
+    # exclude theta
+    idata = get_unconstrained_samples(
+        centered_eight,
+        var_names="~theta",
+    )
+
+    # it returns a datatree
+    assert isinstance(idata, xr.DataTree)
+    dt_out = idata["unconstrained_posterior"]
+    assert "theta" not in dt_out
+    assert "mu" in dt_out
+
+
+def test_with_transform_dimensionality_changed(centered_eight):
+    # transform theta
+    def lagged_cumsum(da):
+        core_dim = da.dims[-1]
+        return da.cumsum(core_dim).isel({core_dim: slice(None, -1)})
+
+    funcs = {"theta": lagged_cumsum}
+
+    idata = get_unconstrained_samples(centered_eight, transform_funcs=funcs)
+
+    dt_post = centered_eight["posterior"]
+    dt_uc = idata["unconstrained_posterior"]
+    # expected values of theta
+    expected_theta = lagged_cumsum(dt_post["theta"])
+    xrt.assert_equal(dt_uc["theta"], expected_theta)
+
+    # assert dims and shape
+    assert dt_uc["theta"].dims == dt_post["theta"].dims
+    assert dt_uc["theta"].coords.keys() == dt_post["theta"].coords.keys()
+    assert dt_uc["theta"].coords["school"].size == dt_post["theta"].coords["school"].size - 1
+    assert dt_uc["theta"].shape[0:2] == dt_post["theta"].shape[0:2]
+    assert dt_uc["theta"].shape[2] == dt_post["theta"].shape[2] - 1
+
+
+def test_with_transform_change_dim_one_var(centered_eight):
+    # transform mu
+    def lagged_cumsum(da):
+        core_dim = da.dims[-1]
+        return da.cumsum(core_dim).isel({core_dim: slice(None, -1)})
+
+    funcs = {"mu": lagged_cumsum}
+
+    idata = get_unconstrained_samples(centered_eight, transform_funcs=funcs)
+
+    dt_post = centered_eight["posterior"]
+    dt_uc = idata["unconstrained_posterior"]
+    # expected values of mu
+    transformed_mu = lagged_cumsum(dt_post["mu"])
+    # fill nan as dimension can't be removed
+    filling_mu = xr.DataArray(np.array([np.nan]), coords={"draw": [9]}, dims="draw")
+    expected_mu = xr.concat([transformed_mu, filling_mu], dim="draw")
```

Comment:
I think this is ok, we can always change it if we end up fining a real case where this happens and is relevant. Code-wise we can simplify this using `pad`. I don't know why I can't add an actual suggestion and I haven't executed this code so some changes might be needed

```python
expected_mu = transformed_mu.pad(draw=(0, 1), constant_values=np.nan)
```

ref: https://docs.xarray.dev/en/latest/generated/xarray.DataArray.pad.html

Reply:
`pad` is really what it's being done, I just had to assign the coords as they are fill with np.nan.